From 569267688d97f263c4a2b69437d26078084b5031 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Tue, 24 Jun 2014 18:38:12 -0400
Subject: [scsi] More USB deadlock fixes

Message-id: <1403635094-26187-5-git-send-email-emilne@redhat.com>
Patchwork-id: 84681
O-Subject: [RHEL7.0 PATCH 4/6] [SCSI] More USB deadlock fixes
Bugzilla: 1103881
RH-Acked-by: Bryn M. Reeves <bmr@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>

This patch fixes a corner case in the previous USB Deadlock fix patch (12023e7
[SCSI] Fix USB deadlock caused by SCSI error handling).

The scenario is abort command, set flag, abort completes, send TUR, TUR
doesn't return, so we now try to abort the TUR, but scsi_abort_eh_cmnd()
will skip the abort because the flag is set and move straight to reset.

Reviewed-by: Hannes Reinecke <hare@suse.de>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
(cherry picked from commit c69e6f812bab0d5442b40e2f1bfbca48d40bc50b)
Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_error.c b/drivers/scsi/scsi_error.c
index 8c62ae2..2be16e8 100644
--- a/drivers/scsi/scsi_error.c
+++ b/drivers/scsi/scsi_error.c
@@ -921,6 +921,7 @@ void scsi_eh_prep_cmnd(struct scsi_cmnd *scmd, struct scsi_eh_save *ses,
  ses->prot_op = scmd->prot_op;
 
  scmd->prot_op = SCSI_PROT_NORMAL;
+ scmd->eh_eflags = 0;
  scmd->cmnd = ses->eh_cmnd;
  memset(scmd->cmnd, 0, BLK_MAX_CDB);
  memset(&scmd->sdb, 0, sizeof(scmd->sdb));
-- 
1.7.1